engine/installation/linux/ubuntulinux.md

```diff
@@ -116,365 +119,494 @@ packages from the Docker repository:
+`/etc/resolv.conf`. The `dnsmasq` service provides a local DNS cache to speed up
+DNS look-ups and also provides DHCP services. However, Docker needs access to a
+fully-functional DNS server. If it detects that no DNS server referenced in
+`/etc/resolv.conf` is a fully functional DNS server, the following warning
+occurs and Docker uses the public DNS servers provided by Google at `8.8.8.8`
+and `8.8.4.4` for DNS resolution.
 
-### Configure a DNS server for use by Docker
+```none
+WARNING: Local (127.0.0.1) DNS resolver found in resolv.conf and containers
+can't use it. Using default external servers : [8.8.8.8 8.8.4.4]
+```
 
-Systems that run Ubuntu or an Ubuntu derivative on the desktop typically use
-`127.0.0.1` as the default `nameserver` in `/etc/resolv.conf` file. The
-NetworkManager also sets up `dnsmasq` to use the real DNS servers of the
-connection and sets up `nameserver 127.0.0.1` in /`etc/resolv.conf`.
+If you don't use `dnsmasq` or NetworkManaager or have never seen this warning,
+you can skip the rest of this section. To see if you use `dnsmasq`, use the
+following command:
 
-When starting containers on desktop machines with these configurations, Docker
-users see this warning:
+```bash
+$ ps aux |grep dnsmasq
+```
 
-    WARNING: Local (127.0.0.1) DNS resolver found in resolv.conf and containers
-    can't use it. Using default external servers : [8.8.8.8 8.8.4.4]
+If this warning occurs and cannot use the public nameservers, such as when you
+run a DNS server which resolves hostnames on your internal network, you have
+two choices:
 
-The warning occurs because Docker containers can't use the local DNS nameserver.
-Instead, Docker defaults to using an external nameserver.
+- You can specify a DNS server for Docker to use.
+- You can disable `dnsmasq` in NetworkManager. If you do this, NetworkManager
+  will add your true DNS nameserver to `/etc/resolv.conf`, but you will lose the
+  possible benefits of `dnsmasq`.
 
-To avoid this warning, you can specify a DNS server for use by Docker
-containers. Or, you can disable `dnsmasq` in NetworkManager. Though, disabling
-`dnsmasq` might make DNS resolution slower on some networks.
+**You only need to use one of these methods.**
 
-The instructions below describe how to configure the Docker daemon
-running on Ubuntu 14.10 or below. Ubuntu 15.04 and above use `systemd`
-as the boot and service manager. Refer to [control and configure Docker
-with systemd](../../admin/systemd.md#custom-docker-daemon-options) to
-configure a daemon controlled by `systemd`.
+#### Specify DNS servers for Docker
 
-To specify a DNS server for use by Docker:
+The instructions below work whether your Ubuntu installation uses `upstart` or
+`systemd`.
 
-1. Log into Ubuntu as a user with `sudo` privileges.
+1.  Log into Ubuntu as a user with `sudo` privileges.
 
-2. Open the `/etc/default/docker` file for editing.
+2.  Check to see if the Docker daemon has been started using a `--config-file`
```

Comment:
Wondering if we should begin with the "possible" exceptions,
or assume the default configuration is used. Maybe we should
"swap" this; something like

```Markdown
The default location of the configuration file is `/etc/docker/daemon.json`.
You can change the location of the configuration file using the `--config-file`
daemon flag. The documentation below assumes the configuration file is
located at `/etc/docker/daemon.json`.
```